File: forms/neurodiversity-adjustment-review/front-end-with-svelte/src/lib/engine/completeness-rules.ts
import type { NeurodiversityAdjustmentReview, FiredRule } from './types';
import { anyEffectivenessAnswered } from './utils';

/** A weighted mandatory section, used to compute Axis C completeness. */
interface SectionCheck {
	ruleId: string;
	weight: number;
	label: string;
	present: (r: NeurodiversityAdjustmentReview) => boolean;
}

/**
 * The mandatory review sections and their weights, per the engine spec. The
 * effectiveness ratings and the worker's own feedback carry the most weight;
 * identification and dating carry the least. Total weight 17.
 */
const sections: SectionCheck[] = [
	{
		ruleId: 'R-COMPLETE-EFFECTIVENESS',
		weight: 3,
		label: 'effectiveness ratings',
		present: (r) => anyEffectivenessAnswered(r)
	},
	{
		ruleId: 'R-COMPLETE-WORKER-FEEDBACK',
		weight: 3,
		label: 'worker feedback',
		present: (r) => r.workerFeedback.trim() !== ''
	},
	{
		ruleId: 'R-COMPLETE-SATISFACTION',
		weight: 2,
		label: 'worker satisfaction',
		present: (r) => r.workerSatisfied !== ''
	},
	{
		ruleId: 'R-COMPLETE-WELLBEING',
		weight: 2,
		label: 'wellbeing change',
		present: (r) => r.wellbeingChange !== ''
	},
	{
		ruleId: 'R-COMPLETE-NEXT-REVIEW',
		weight: 2,
		label: 'next review date',
		present: (r) => r.nextReviewDate.trim() !== ''
	},
	{
		ruleId: 'R-COMPLETE-REVIEW-DATE',
		weight: 1,
		label: 'review date',
		present: (r) => r.reviewDate.trim() !== ''
	},
	{
		ruleId: 'R-COMPLETE-METHOD',
		weight: 1,
		label: 'review method',
		present: (r) => r.reviewMethod !== ''
	},
	{
		ruleId: 'R-COMPLETE-WORKER-NAME',
		weight: 1,
		label: 'worker name',
		present: (r) => r.workerName.trim() !== ''
	},
	{
		ruleId: 'R-COMPLETE-MANAGER',
		weight: 1,
		label: 'manager / HR contact',
		present: (r) => r.managerName.trim() !== ''
	},
	{
		ruleId: 'R-COMPLETE-STATUS',
		weight: 1,
		label: 'review status',
		present: (r) => r.reviewStatus !== ''
	}
];

const TOTAL_WEIGHT = sections.reduce((sum, s) => sum + s.weight, 0);

/**
 * Axis C — review completeness.
 *
 * Returns the weighted percentage (0–100, rounded) of mandatory sections that
 * are present, plus an audit-trail rule for each missing section.
 */
export function gradeCompleteness(r: NeurodiversityAdjustmentReview): {
	completenessPercent: number;
	firedRules: FiredRule[];
} {
	const firedRules: FiredRule[] = [];
	let presentWeight = 0;

	for (const section of sections) {
		if (section.present(r)) {
			presentWeight += section.weight;
		} else {
			firedRules.push({
				ruleId: section.ruleId,
				axis: 'completeness',
				category: 'missing-field',
				description: `Mandatory review section missing: ${section.label}.`
			});
		}
	}

	const completenessPercent = Math.round((presentWeight / TOTAL_WEIGHT) * 100);
	return { completenessPercent, firedRules };
}

File: forms/neurodiversity-adjustment-review/back-end-with-loco/src/engine/completeness_rules.rs
//! Axis C — review completeness.
//!
//! Returns the weighted percentage (0–100, rounded) of mandatory review
//! sections that are present, plus an audit-trail rule for each missing
//! section. Weights reflect how central each section is to a valid ACAS
//! reasonable-adjustments review record.

use super::types::{FiredRule, NeurodiversityAdjustmentReview};
use super::utils::any_effectiveness_answered;

struct SectionCheck {
    weight: i32,
    rule_id: &'static str,
    label: &'static str,
    present: fn(&NeurodiversityAdjustmentReview) -> bool,
}

const SECTIONS: &[SectionCheck] = &[
    SectionCheck {
        weight: 3,
        rule_id: "R-COMPLETE-EFFECTIVENESS",
        label: "effectiveness ratings",
        present: any_effectiveness_answered,
    },
    SectionCheck {
        weight: 3,
        rule_id: "R-COMPLETE-WORKER-FEEDBACK",
        label: "worker feedback",
        present: |r| !r.worker_feedback.trim().is_empty(),
    },
    SectionCheck {
        weight: 2,
        rule_id: "R-COMPLETE-SATISFACTION",
        label: "worker satisfaction",
        present: |r| !r.worker_satisfied.is_empty(),
    },
    SectionCheck {
        weight: 2,
        rule_id: "R-COMPLETE-WELLBEING",
        label: "wellbeing change",
        present: |r| !r.wellbeing_change.is_empty(),
    },
    SectionCheck {
        weight: 2,
        rule_id: "R-COMPLETE-NEXT-REVIEW",
        label: "next review date",
        present: |r| !r.next_review_date.trim().is_empty(),
    },
    SectionCheck {
        weight: 1,
        rule_id: "R-COMPLETE-REVIEW-DATE",
        label: "review date",
        present: |r| !r.review_date.trim().is_empty(),
    },
    SectionCheck {
        weight: 1,
        rule_id: "R-COMPLETE-METHOD",
        label: "review method",
        present: |r| !r.review_method.is_empty(),
    },
    SectionCheck {
        weight: 1,
        rule_id: "R-COMPLETE-WORKER-NAME",
        label: "worker name",
        present: |r| !r.worker_name.trim().is_empty(),
    },
    SectionCheck {
        weight: 1,
        rule_id: "R-COMPLETE-MANAGER",
        label: "manager / HR contact",
        present: |r| !r.manager_name.trim().is_empty(),
    },
    SectionCheck {
        weight: 1,
        rule_id: "R-COMPLETE-STATUS",
        label: "review status",
        present: |r| !r.review_status.is_empty(),
    },
];

/// Grade completeness, returning the percent and one fired rule per missing
/// section.
#[must_use]
pub fn grade_completeness(r: &NeurodiversityAdjustmentReview) -> (i32, Vec<FiredRule>) {
    let mut fired_rules: Vec<FiredRule> = Vec::new();
    let mut present_weight = 0_i32;
    let mut total_weight = 0_i32;

    for section in SECTIONS {
        total_weight += section.weight;
        if (section.present)(r) {
            present_weight += section.weight;
        } else {
            fired_rules.push(FiredRule {
                rule_id: section.rule_id.to_string(),
                axis: "completeness".to_string(),
                category: "missing-field".to_string(),
                description: format!("Mandatory review section missing: {}.", section.label),
            });
        }
    }

    let completeness_percent = if total_weight == 0 {
        0
    } else {
        ((f64::from(present_weight) / f64::from(total_weight)) * 100.0).round() as i32
    };
    (completeness_percent, fired_rules)
}
